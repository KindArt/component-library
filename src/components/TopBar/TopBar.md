Standard:

```tsx
const links = [
  { title: 'Link 1', url: '/link-1', order: 0 },
  { title: 'Link 2', url: '/link-2', order: 0 },
  { title: 'Link 3', url: '/link-3', order: 0 },
];
<TopBar title="Title" links={links} />;
```
