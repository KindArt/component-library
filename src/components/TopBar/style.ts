import styled, { css } from 'styled-components';
import { MediaQueries, Typography, Colours } from '../../constants/styles';

interface ITopBarProps {
  leftAlign?: boolean;
}

export const BurgerMenuStyle = css`
  display: none;

  @media (max-width: 768px) {
    display: block;
    border: none !important;
  }
`;

export const TopBarStyle = styled.div<ITopBarProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  color: ${({ theme }) => theme.topBar.textColour};
  background-color: ${({ theme }) => theme.topBar.backgroundColour};
  height: 70px;

  ${({ className }) =>
    className &&
    css`
      ${className}
    `}
`;

TopBarStyle.defaultProps = {
  theme: {
    topBar: {
      backgroundColour: '#252529',
      textColour: '#FFF',
    },
  },
};

export const Container = styled.div<ITopBarProps>`
  display: flex;
  justify-content: space-between;
  align-items: stretch;
  width: ${MediaQueries.mobileContainerWidth};

  @media (min-width: ${MediaQueries.tabletBreakpointWidth}) {
    width: ${MediaQueries.tabletContainerWidth};
  }

  @media (min-width: ${MediaQueries.desktopBreakpointWidth}) {
    width: ${MediaQueries.desktopContainerWidth};
  }

  ${({ leftAlign }) =>
    leftAlign &&
    css`
      justify-content: left;
    `}
`;

export const Brand = styled.div<ITopBarProps>`
  display: flex;
  align-items: center;
  font-size: 1.25rem;

  ${({ leftAlign }) =>
    leftAlign &&
    css`
      margin-right: 20px;
    `}
`;

export const Links = styled.div`
  display: none;
  font-size: ${Typography.DEFAULT_FONT_SIZE};

  @media (min-width: ${MediaQueries.tabletBreakpointWidth}) {
    display: flex;
    align-items: stretch;
  }

  ${({ className }) =>
    className &&
    css`
      ${className}
    `}
`;

export const Actions = styled.div<ITopBarProps>`
  display: flex;
  align-items: center;

  ${({ leftAlign }) =>
    leftAlign &&
    css`
      margin-left: auto;
    `}
`;

export const Link = styled.div`
  display: flex;
  align-items: center;
  cursor: pointer;

  :not(:last-child) {
    padding-right: 10px;
  }

  :hover {
    color: ${Colours.TEXT_SECONDARY};
  }
`;
