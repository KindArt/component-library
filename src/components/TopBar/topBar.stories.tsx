import React from 'react';
import TopBar from '.';

export default {
  component: TopBar,
  title: 'Top Bar'
};

const customNavigationMock = (url: string) => console.log('The url is ', url);

const defaultProps = {
  toggleMenu: () => {},
  isMenuOpen: false,
  title: 'Title',
  links: [
    { title: 'Home', url: '#', order: 1 },
    { title: 'About', url: '#', order: 2, customNavigation: customNavigationMock }
  ]
};

export const normal = () => <TopBar {...defaultProps}></TopBar>;
