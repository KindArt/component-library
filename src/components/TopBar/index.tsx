import React, { FC, ReactNode } from 'react';
import { Button, BurgerMenu, MobileSidebar } from '../';
import { BurgerMenuStyle, TopBarStyle, Container, Brand, Links, Actions, Link } from './style';
import { ILinks } from 'types';

export interface ITopBarProps {
  brand?: ReactNode;
  links?: ILinks[];
  actions?: ReactNode;
  leftAlign?: boolean;
  toggleMenu?: Function;
  isMenuOpen?: boolean;
  classOverrides?: {
    wrapper?: string;
    linkContainer?: string;
  };
}

const renderLinks: Function = (links: ILinks[]) => {
  const sortedLinks = links.sort((a, b) => {
    if (!a.order || !b.order) {
      return 0;
    }
    return a.order - b.order;
  });

  return sortedLinks.map(({ title, path, navigate }, index) => {
    return (
      <Link key={index} onClick={() => navigate(path)}>
        {title}
      </Link>
    );
  });
};

const TopBar: FC<ITopBarProps> = ({
  brand,
  links,
  actions,
  leftAlign,
  toggleMenu,
  isMenuOpen = false,
  classOverrides,
}) => {
  return (
    <TopBarStyle className={classOverrides && classOverrides.wrapper}>
      {toggleMenu && (
        <Button transparent clear classOverrides={{ buttonWrapper: `${BurgerMenuStyle}` }} onClick={() => toggleMenu()}>
          <BurgerMenu isOpen={isMenuOpen}></BurgerMenu>
        </Button>
      )}
      <Container leftAlign={leftAlign}>
        {brand && (
          <Brand data-testid="TopBar-brand" leftAlign={leftAlign}>
            {brand}
          </Brand>
        )}
        {links && (
          <Links data-testid="TopBar-links" className={classOverrides && classOverrides.linkContainer}>
            {renderLinks(links)}
          </Links>
        )}
        {actions && (
          <Actions data-testid="TopBar-actions" leftAlign={leftAlign}>
            {actions}
          </Actions>
        )}
      </Container>
      {links && <MobileSidebar isVisible={isMenuOpen} closeMenu={toggleMenu} links={links} />}
    </TopBarStyle>
  );
};

export default TopBar;
