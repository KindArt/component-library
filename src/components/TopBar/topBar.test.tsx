import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import TopBar, { ITopBarProps } from '.';

describe('<TopBar />', () => {
  let toggleMenuMock = jest.fn();

  const getTopBar = (props?: any) => {
    const defaultProps: ITopBarProps = {
      toggleMenu: toggleMenuMock,
      isMenuOpen: false,
      ...props,
    };
    const utils = render(<TopBar {...defaultProps} />);
    return { ...utils };
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    const { container } = getTopBar();
    expect(container).toMatchSnapshot();
  });

  describe('Visuals', () => {
    it('should render the branding if it is passed', () => {
      const props = { brand: 'Branding' };
      const { queryByText } = getTopBar(props);
      const brandText = queryByText('Branding');
      expect(brandText).toBeTruthy();
    });

    it('should not render the branding if it is not passed', () => {
      const { queryByTestId } = getTopBar();
      const brand = queryByTestId('TopBar-brand');
      expect(brand).toBeFalsy();
    });

    it('should render links if they are passed', () => {
      const mockNavigate = jest.fn();
      const props = { links: [{ title: 'Link 1', url: '/', navigate: mockNavigate }] };
      const { queryAllByText } = getTopBar(props);
      const link = queryAllByText('Link 1');
      expect(link).toHaveLength(2);
    });

    it('should not render the links if it is not passed', () => {
      const { queryByTestId } = getTopBar();
      const links = queryByTestId('TopBar-links');
      expect(links).toBeFalsy();
    });

    it('should render the actions if they are passed', () => {
      const props = { actions: 'Action' };
      const { queryByText } = getTopBar(props);
      const action = queryByText('Action');
      expect(action).toBeTruthy();
    });

    it('should not render the links if it is not passed', () => {
      const { queryByTestId } = getTopBar();
      const actions = queryByTestId('TopBar-actions');
      expect(actions).toBeFalsy();
    });

    it('should render the burger menu if toggleMenu is passed', () => {
      const { queryByTestId } = getTopBar();
      const burgerMenu = queryByTestId('Button-button');
      expect(burgerMenu).toBeTruthy();
    });

    it('should not render the burger menu if toggleMenu is not passed', () => {
      const { queryByTestId } = getTopBar({ toggleMenu: null });
      const burgerMenu = queryByTestId('Button-button');
      expect(burgerMenu).toBeFalsy();
    });
  });

  describe('Actions', () => {
    it('should call navigate on click', async () => {
      const mockNavigate = jest.fn();
      const props = { links: [{ title: 'Link 1', url: '/', navigate: mockNavigate }] };
      const { getAllByText } = getTopBar(props);
      const link = getAllByText('Link 1')[0];

      await act(async () => {
        await fireEvent.click(link);
      });

      expect(mockNavigate).toHaveBeenCalledTimes(1);
    });
  });
});
