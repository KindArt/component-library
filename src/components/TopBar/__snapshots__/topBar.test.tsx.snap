// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar /> should render 1`] = `
<div>
  <div
    class="sc-fzoYkl fCdOVl"
  >
    <button
      class="sc-AxirZ leDdxt 
  display: none;

  @media (max-width: 768px) {
    display: block;
    border: none !important;
  }
"
      data-testid="Button-button"
      type="button"
    >
      <span
        class="sc-AxiKw cGEoKA"
        data-testid="Button-content"
      >
        <div
          class="sc-fzqARJ hkvVQj"
        >
          <svg
            class="sc-fzqNqU fcljRp"
            fill="currentColor"
            height="2px"
            version="1.1"
            viewBox="0 0 20 2"
            width="20px"
          >
            <g
              id="Icon/20px/menu-[Extra]"
              transform="translate(0.000000, -9.000000)"
            >
              <polygon
                id="Path"
                points="0 11 20 11 20 9 0 9"
              />
            </g>
          </svg>
          <svg
            class="sc-fzoyTs kCGEHg"
            fill="currentColor"
            height="2px"
            version="1.1"
            viewBox="0 0 20 2"
            width="20px"
          >
            <g
              id="Icon/20px/menu-[Extra]"
              transform="translate(0.000000, -9.000000)"
            >
              <polygon
                id="Path"
                points="0 11 20 11 20 9 0 9"
              />
            </g>
          </svg>
          <svg
            class="sc-fzoNJl bnuJxw"
            fill="currentColor"
            height="2px"
            version="1.1"
            viewBox="0 0 20 2"
            width="20px"
          >
            <g
              id="Icon/20px/menu-[Extra]"
              transform="translate(0.000000, -9.000000)"
            >
              <polygon
                id="Path"
                points="0 11 20 11 20 9 0 9"
              />
            </g>
          </svg>
        </div>
      </span>
    </button>
    <div
      class="sc-fzpkJw fwQomI"
    />
  </div>
</div>
`;
